cpp/src/arrow/util/io_util.cc

```diff
@@ -1867,5 +1889,42 @@ uint64_t GetOptionalThreadId() {
   return (tid == 0) ? tid - 1 : tid;
 }
 
+// Returns the current resident set size (physical memory use) measured
+// in bytes, or zero if the value cannot be determined on this OS.
+// See: https://stackoverflow.com/a/14927379
+int64_t GetCurrentRSS() {
+#if defined(_WIN32)
+  // Windows --------------------------------------------------
+  PROCESS_MEMORY_COUNTERS info;
+  GetProcessMemoryInfo(GetCurrentProcess(), &info, sizeof(info));
+  return (int64_t)info.WorkingSetSize;
```

Comment:
```suggestion
  return static_cast<int64_t>(info.WorkingSetSize);
```
For consistency with the rest of the codebase here (and elsewhere in this file) we should prefer C++ style casts: